### PR TITLE
Migrate DefaultEntityReference

### LINKED ISSUE #992: Migrate `defaultEntityReference` to C++
## What

Migrate `defaultEntityReference` to C++ using the callback pattern (back-end API requests may be required), and updated return types defined for when there is no suitable reference.

## Why

- C++ implementation needed.
- The previous approach of using an empty string is not valid, now that we have a strong-type for EntityReference that is assumed to be valid for other API methods.

## ACs

- Removed from python, and implemented in C++ using the callback pattern.
- Value type is `EntityReference`,  `kEntityAccessErrorBEE` when one isn't available, `kInvalidTraitSetBEE` if the set is not known to the manager.
- A default implementation that returns the `kEntityAccessErrorBEE` for each input so the method is optional for managers.
- API compliance for a later date

### LINKED ISSUE #995: Remove Python`Manager`/`ManagerFactory`
## What

Remove `Manager`/`ManagerFactory`, and make C++ classes final.

## ACs

- File removed
- Ensure hoist in Python module

### PR DESCRIPTION
**Pull request by elliotcmorris**
## Description
Migrate DefaultEntityReference from python to C++, migrating to the callback pattern.

Adds `kInvalidTraitSet` and equivalent exception for usage by managers in this, and other methods.

Also removes the `Manager.py` and `ManagerFactory.py`, mostly because keeping them around was making things difficult, full removal turned out easier

Closes #992 
Closes #995

- [x] I have updated the release notes.
- [x] I have updated all relevant user documentation.

## Reviewer notes
The first commit of this thing relates to a tangential thing with the `preflightHint` error, since it had been left out of some places where we expect and want to have the full list of error/exceptions. Noticed when adding the new exception, and thought best just do that while I was there.

Also, I _think_ that removing the python files is a transparent change, but I'm not certain. I haven't added anything about that to release notes yet because that is my presumption.

## Why is this a draft?
Because it strikes me that it might be better to have a commit between migrating the function and deleting the python files wholesale, so there may be some git history work to do next week.
I also want to check in with folks about the `__repr__` and `__str__` functionality that was "lost", and double check docstring content hasn't been lost in the full file delete.
However, since i'm travelling all tuesday morning, figured i'd get this up anyhow.

### CODE REVIEW

File: RELEASE_NOTES.md [hunk 1]
```diff
@@ -39,13 +39,21 @@ v1.0.0-alpha.x
...
+  API. [#992](https://github.com/OpenAssetIO/OpenAssetIO/issues/992)
+
 ### New Features
 
 - Added `Context.Access.kCreateRelated` access pattern, to  indicate
   when a workflow specifically creates a new entity as a relation to an
   existing one.
   [1016](https://github.com/OpenAssetIO/OpenAssetIO/issues/1016)
 
+- Added `BatchElementError.ErrorCode.InvalidTraitSet` and
```
Comment (feltech) [inline RELEASE_NOTES.md:53]: [fix] 
```suggestion
- Added `BatchElementError.ErrorCode.kInvalidTraitSet` and
```

File: doc/doxygen/src/MainPage.dox [hunk 2]
```diff
@@ -170,12 +170,11 @@
...
  * one of the two abstract interfaces defined by the API, illustrated in
  * orange above - either the @fqref{hostApi.HostInterface}
- * "HostInterface" or @ref
- * openassetio.managerApi.ManagerInterface.ManagerInterface
+ * "HostInterface" or @fqref{managerApi.ManagerInterface}
  * "ManagerInterface".
  *
  * - The asset management system implements the
- *   @ref openassetio.managerApi.ManagerInterface.ManagerInterface
+ *   @fqref{managerApi.ManagerInterface}.ManagerInterface
```
Comment (feltech) [inline doc/doxygen/src/MainPage.dox:177]: [consider][minor]
```suggestion
 *   @fqref{managerApi.ManagerInterface}. ManagerInterface
```

**Review by feltech — COMMENTED** [review]
Minor typos from WIP review, looking good so far.

File: doc/doxygen/src/ForManagers.dox [hunk 3]
```diff
@@ -117,7 +117,7 @@
  *
  * - Implement the methods grouped under the "Entity Reference Inspection"
  *   and "Entity Reference Resolution" sections of the
- *   @ref openassetio.managerApi.ManagerInterface.ManagerInterface
+ *   @fqref{managerApi.ManagerInterface}.ManagerInterface
```
Comment (feltech) [inline doc/doxygen/src/ForManagers.dox:120]: [fix] This links to the `ManagerInterface` constructor, probably not what was intended.

File: doc/doxygen/src/ForManagers.dox [hunk 4]
```diff
@@ -135,7 +135,7 @@
  *
  * - Implement the methods grouped under the "Publishing" section, and
  *   the remaining methods in the "Entity Information" section of
- *   @ref openassetio.managerApi.ManagerInterface.ManagerInterface
+ *   @fqref{managerApi.ManagerInterface}.ManagerInterface
```
Comment (feltech) [inline doc/doxygen/src/ForManagers.dox:138]: [fix] This links to the `ManagerInterface` constructor, probably not what was intended.

File: doc/doxygen/src/ForManagers.dox [hunk 5]
```diff
@@ -197,7 +197,7 @@
  * In order to support entity relationships:
  *
  * - Implement the API methods grouped under
- *   @ref openassetio.managerApi.ManagerInterface.ManagerInterface
+ *   @fqref{managerApi.ManagerInterface}.ManagerInterface
```
Comment (feltech) [inline doc/doxygen/src/ForManagers.dox:200]: [fix] This links to the `ManagerInterface` constructor, probably not what was intended.

File: doc/doxygen/src/Glossary.dox [hunk 6]
```diff
@@ -180,13 +180,16 @@
...
+ * a host and an @ref asset_management_system in order to persist or 
+ * query data.
+ * A host communicates with a @ref manager via the
+ * @fqref{hostApi.Manager} "Manager" class.
  *
  * > Note: The Manager class is an abstraction, that takes care of
  * > sundry housekeeping and helps manage session state. Internally it
- * > holds an instance of the @ref openassetio.managerApi.ManagerInterface.ManagerInterface
+ * > holds an instance of the
+ * > @fqref{managerApi.ManagerInterface}.ManagerInterface
```
Comment (feltech) [inline doc/doxygen/src/Glossary.dox:192]: [fix] This links to the `ManagerInterface` constructor, probably not what was intended.

File: doc/doxygen/src/Glossary.dox [hunk 7]
```diff
@@ -232,7 +235,7 @@
  *
  * @section ManagerInterface
  *
- * The @ref openassetio.managerApi.ManagerInterface.ManagerInterface
+ * The @fqref{managerApi.ManagerInterface}.ManagerInterface
```
Comment (feltech) [inline doc/doxygen/src/Glossary.dox:238]: [fix] This links to the `ManagerInterface` constructor, probably not what was intended.

File: src/openassetio-core/include/openassetio/hostApi/Manager.hpp [hunk 8]
```diff
@@ -582,15 +582,15 @@ class OPENASSETIO_CORE_EXPORT Manager {
    * corresponding index of the entity reference in @p entityReferences
    * along with a boolean indicating existence, as defined above. The
    * callback will be called on the same thread that initiated the call
-   * to `resolve`.
+   * to `entityExists`.
```
Comment (feltech) [inline src/openassetio-core/include/openassetio/hostApi/Manager.hpp:585]: [consider][minor] Good spot, ideally would be a separate commit, but probably not worth the effort.

File: src/openassetio-core/include/openassetio/hostApi/Manager.hpp [hunk 9]
```diff
@@ -844,6 +844,51 @@ class OPENASSETIO_CORE_EXPORT Manager {
...
       const ContextConstPtr& context, const BatchElementErrorPolicyTag::Variant& errorPolicyTag);
 
+  /**
+   * Callback signature used for a successful default entity reference query.
+   */
+  using DefaultEntityReferenceSuccessCallback = std::function<void(std::size_t, EntityReference)>;
+
+  /**
+   * Called to determine an @ref EntityReference considered to be a
+   * sensible default for each of the given entity @ref trait_set "trait sets"
```
Comment (feltech) [inline src/openassetio-core/include/openassetio/hostApi/Manager.hpp:854]: [consider][minor] Line wrapping

File: src/openassetio-core/include/openassetio/managerApi/ManagerInterface.hpp [hunk 10]
```diff
@@ -790,6 +790,68 @@ class OPENASSETIO_CORE_EXPORT ManagerInterface {
...
+   * situation.
+   *
+   * For example, a host may request the default ref for the @ref
+   * trait_set of a 'ShotSpecification' with access 'kWrite'.
+   * If the Manager has some concept of the 'current sequence' it may
+   * wish to return this so that a 'Create Shots' action starts
+   * somewhere meaningful.
+   *
+   * The default implementation returns a @fqref{BatchElementError.ErrorCode.kEntityAccessError}
+   * "kEntityAccessError" indicating that no valid default exists.`
```
Comment (feltech) [inline src/openassetio-core/include/openassetio/managerApi/ManagerInterface.hpp:818]: [fix] Errant ``` ` ```

**Review by feltech — CHANGES_REQUESTED** [review]
More WIP minor review comments

File: src/openassetio-python/package/openassetio/pluginSystem/PythonPluginSystemManagerPlugin.py [hunk 11]
```diff
@@ -64,16 +64,17 @@ def identifier():
 
         @return str
 
-        @see @ref openassetio.managerApi.ManagerInterface
+        @see @fqref{managerApi.ManagerInterface}
```
Comment (feltech) [inline src/openassetio-python/package/openassetio/pluginSystem/PythonPluginSystemManagerPlugin.py:67]: [consider][minor] Line wrapping

File: src/openassetio-python/package/openassetio/pluginSystem/PythonPluginSystemManagerPlugin.py [hunk 12]
```diff
@@ -64,16 +64,17 @@ def identifier():
...
         "ManagerInterface"
         """
         raise NotImplementedError
 
     @classmethod
     def interface(cls):
         """
-        Constructs an instance of the @ref
-        openassetio.managerApi.ManagerInterface.
+        Constructs an instance of the
```
Comment (feltech) [inline src/openassetio-python/package/openassetio/pluginSystem/PythonPluginSystemManagerPlugin.py:75]: [consider][minor] Line wrapping

File: src/openassetio-python/tests/package/managerApi/test_managerinterface.py [hunk 13]
```diff
@@ -152,19 +152,51 @@ def test_default_implementation_raises_RuntimeError(self, a_host_session):
...
+    def test_method_defined_in_cpp(self, method_introspector):
+        assert not method_introspector.is_defined_in_python(
+            ManagerInterface.defaultEntityReference
+        )
         assert method_introspector.is_implemented_once(ManagerInterface, "defaultEntityReference")
 
-    def test_when_given_single_trait_set_then_returns_single_empty_ref(self, manager_interface):
-        refs = manager_interface.defaultEntityReference([()], Mock(), Mock())
-        assert refs == [""]
+    def test_when_given_single_trait_set_then_calls_error_once_for_each_trait_set(
```
Comment (feltech) [inline src/openassetio-python/tests/package/managerApi/test_managerinterface.py:161]: [consider][minor] Just "calls_error_once"?

File: RELEASE_NOTES.md (hunk 1 above)
Comment (foundrytom) [reply RELEASE_NOTES.md]: https://github.com/OpenAssetIO/OpenAssetIO/pull/1058/commits/b0d82f577c00c4ef02e74f799af5b8ef68dadc7f

File: doc/doxygen/src/Glossary.dox (hunk 7 above)
Comment (foundrytom) [reply doc/doxygen/src/Glossary.dox]: https://github.com/OpenAssetIO/OpenAssetIO/pull/1058/commits/948efe5518f4da207ec3438f17531c456b524965

File: doc/doxygen/src/Glossary.dox (hunk 6 above)
Comment (foundrytom) [reply doc/doxygen/src/Glossary.dox]: https://github.com/OpenAssetIO/OpenAssetIO/pull/1058/commits/948efe5518f4da207ec3438f17531c456b524965

File: doc/doxygen/src/ForManagers.dox (hunk 5 above)
Comment (foundrytom) [reply doc/doxygen/src/ForManagers.dox]: https://github.com/OpenAssetIO/OpenAssetIO/pull/1058/commits/948efe5518f4da207ec3438f17531c456b524965

File: doc/doxygen/src/ForManagers.dox (hunk 4 above)
Comment (foundrytom) [reply doc/doxygen/src/ForManagers.dox]: https://github.com/OpenAssetIO/OpenAssetIO/pull/1058/commits/948efe5518f4da207ec3438f17531c456b524965

File: doc/doxygen/src/ForManagers.dox (hunk 3 above)
Comment (foundrytom) [reply doc/doxygen/src/ForManagers.dox]: https://github.com/OpenAssetIO/OpenAssetIO/pull/1058/commits/948efe5518f4da207ec3438f17531c456b524965

File: doc/doxygen/src/MainPage.dox (hunk 2 above)
Comment (foundrytom) [reply doc/doxygen/src/MainPage.dox]: https://github.com/OpenAssetIO/OpenAssetIO/pull/1058/commits/948efe5518f4da207ec3438f17531c456b524965

File: src/openassetio-core/include/openassetio/managerApi/ManagerInterface.hpp (hunk 10 above)
Comment (foundrytom) [reply src/openassetio-core/include/openassetio/managerApi/ManagerInterface.hpp]: https://github.com/OpenAssetIO/OpenAssetIO/pull/1058/commits/44ca15e5346d5ee9e7d1a3aaa30e0636a0f94aae

File: src/openassetio-core/include/openassetio/hostApi/Manager.hpp (hunk 9 above)
Comment (foundrytom) [reply src/openassetio-core/include/openassetio/hostApi/Manager.hpp]: https://github.com/OpenAssetIO/OpenAssetIO/pull/1058/commits/44ca15e5346d5ee9e7d1a3aaa30e0636a0f94aae

**Review by feltech — APPROVED** [review]

File: src/openassetio-python/package/openassetio/pluginSystem/PythonPluginSystemManagerPlugin.py (hunk 12 above)
Comment (foundrytom) [reply src/openassetio-python/package/openassetio/pluginSystem/PythonPluginSystemManagerPlugin.py:74]: https://github.com/OpenAssetIO/OpenAssetIO/pull/1058/commits/4c012fff18b7da2119be2eda72cc0c8a614e2335

File: src/openassetio-python/package/openassetio/pluginSystem/PythonPluginSystemManagerPlugin.py (hunk 11 above)
Comment (foundrytom) [reply src/openassetio-python/package/openassetio/pluginSystem/PythonPluginSystemManagerPlugin.py]: https://github.com/OpenAssetIO/OpenAssetIO/pull/1058/commits/4c012fff18b7da2119be2eda72cc0c8a614e2335

File: src/openassetio-python/tests/package/managerApi/test_managerinterface.py (hunk 13 above)
Comment (foundrytom) [reply src/openassetio-python/tests/package/managerApi/test_managerinterface.py]: https://github.com/OpenAssetIO/OpenAssetIO/pull/1058/commits/6f033845e6ff93964809866aacd0d316dad09276

File: src/openassetio-core/include/openassetio/hostApi/Manager.hpp (hunk 8 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/hostApi/Manager.hpp:585]: Not worth the effort :+1:

**Review by foundrytom — APPROVED** [review]
Nice one!
